src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.h

```diff
@@ -0,0 +1,942 @@
+        }
+    }
+
+    /**
+     * @brief Signals completion of new value modifications
+     * @return CHIP_NO_ERROR on success
+     * @retval CHIP_ERROR_INCORRECT_STATE if not in kInitialized state
+     */
+    CHIP_ERROR MarkAsAssigned()
+    {
+        if (mUpdateState != UpdateState::kInitialized)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+        mUpdateState = UpdateState::kAssigned;
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Validates and prepares the new value for commit
+     * @param aUpdCtx Context pointer for callback
+     * @param aUpdCb Callback to invoke on successful commit
+     * @return CHIP_NO_ERROR if validation succeeds
+     * @retval CHIP_ERROR_INCORRECT_STATE if not in kAssigned state
+     */
+    CHIP_ERROR UpdateBegin(void * aUpdCtx, void (*aUpdCb)(uint32_t, void *), bool valid_is_req)
+    {
+        /* Skip if the attribute object has no new attached data */
+        if (mUpdateState == UpdateState::kIdle)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        if (mUpdateState != UpdateState::kAssigned)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+
+        CHIP_ERROR err = CHIP_NO_ERROR;
+
+        // Bypass validation if ctx not set
+        if (valid_is_req && aUpdCtx != nullptr)
+        {
+            mAuxData = aUpdCtx;
+            err      = ValidateNewValue();
+        }
+
+        if (err == CHIP_NO_ERROR)
+        {
+            if (aUpdCb != nullptr)
+            {
+                mAuxCb = aUpdCb;
+            }
+
+            mUpdateState = UpdateState::kValidated;
+        }
+        else
+        {
+            ChipLogError(NotSpecified, "The value of attr %d is not valid!", mAttrId);
+        }
+
+        return err;
+    }
+
+    /**
+     * @brief Commits the validated new value
+     * @note Performs atomic swap if value changed
+     */
+    void UpdateCommit()
+    {
+        /* Skip if the attribute object has no new attached data */
+        if (mUpdateState == UpdateState::kIdle)
+        {
+            return;
+        }
+
+        assert(mUpdateState == UpdateState::kValidated);
+
+        if (HasChanged())
+        {
+            if (mHoldState == StorageState::kHold)
+            {
+                Cleanup(); // Cleanup current value
+            }
+            mHoldState   = StorageState::kHold;
+            mUpdateState = UpdateState::kUpdated;
+        }
+
+        UpdateEnd();
+    }
+
+    /**
+     * @brief Aborts the current update process
+     * @note Performs cleanup and resets to idle state
+     */
+    void UpdateEnd()
+    {
+        /* Skip if the attribute object has no new attached data */
+        if (mUpdateState == UpdateState::kIdle)
+        {
+            return;
+        }
+
+        if (mUpdateState == UpdateState::kUpdated)
+        {
+            SwapActiveValueStorage();
+
+            if (mAuxCb)
+            {
+                mAuxCb(mAttrId, mAuxData);
```

Comment:
So the arg above there was of type AttributeId?  That would have been a good start on making it clear what gets passed to the callback...

Reply:
https://github.com/project-chip/connectedhomeip/issues/40122